runtime/compiler/control/CompilationThread.cpp

```diff
@@ -6028,6 +6033,48 @@ void *TR::CompilationInfo::compileOnSeparateThread(J9VMThread * vmThread, TR::Il
       if (compErrCode)
          *compErrCode = (TR_CompilationErrorCode)entry->_compErrCode;
 
+      // If a remote sync compilation has been changed to a local sync compilation,
+      // queue a new remote async compilation entry for this method.
+      // - The second compilation is queued only when this is the last thread waiting
+      //   for the compilation of the method.
+      // - If for any reason, the current compilation request is changed to a sync
+      //   compilation from an async compilation, it probably means we can’t do an
+      //   async compilation. Don't queue the second async remote compilation.
+      if (entry->_changedToLocalSyncComp &&
+          (entry->_numThreadsWaiting <= 1) &&
+          !forcedSync &&
+          (startPC && startPC != oldStartPC))
+         {
+         acquireCompMonitor(vmThread);
+
+         void *currOldStartPC = startPCIfAlreadyCompiled(vmThread, details, oldStartPC);
+         CompilationPriority compPriority = CP_ASYNC_NORMAL;
+         bool queuedForRemote = false;
+         bool isAsync = true;
+         TR_OptimizationPlan *plan = TR_OptimizationPlan::alloc(entry->_origOptLevel);
+         TR_YesNoMaybe mthInSharedCache = TR_no;
+
+         if (plan)
+            {
+            TR_MethodToBeCompiled *entryRemoteCompReq = addMethodToBeCompiled(details, currOldStartPC, compPriority, isAsync,
+                                                                              plan, &queuedForRemote, mthInSharedCache);
+            if (entryRemoteCompReq)
+               {
+               entryRemoteCompReq->_async = isAsync;
+
+               if (getMethodQueueSize() <= 1 ||
+                  getNumCompThreadsJobless() > 0) // send notification if any thread is sleeping on comp monitor waiting for suitable work
+                  {
+                  getCompilationMonitor()->notifyAll();
+                  }
+               }
+            if ((!queuedForRemote) || (!entryRemoteCompReq))
```

Comment:
The test on queuedForRemote is enough because if `addMethodToBeCompiled` cannot allocate an entry, then it's not going to be queued for sure.